### ISSUE TITLE
Throws when used with JSON columns

### ISSUE THREAD

**Issue by aradalvand** [efcore/EFCore.NamingConventions#234]
Repro:

`Program.cs`:
```csharp
using Microsoft.EntityFrameworkCore;
using Npgsql;

using var db = new AppDbContext();
db.Database.EnsureDeleted();
db.Database.EnsureCreated();

class AppDbContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) =>
        optionsBuilder
            .UseNpgsql("CONNECTION_STRING")
            .UseSnakeCaseNamingConvention() // NOTE: removing this line makes things work, so it's clear that `EFCore.NamingConventions` is the cause of the issue
            .EnableSensitiveDataLogging()
            .LogTo(Console.WriteLine, Microsoft.Extensions.Logging.LogLevel.Information);

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Product>()
            .OwnsOne(p => p.MainTrait, b => b.ToJson())
            .OwnsMany(p => p.Traits, b => b.ToJson());
    }

    public DbSet<Product> Products => Set<Product>();
}

class Product
{
    public int Id { get; set; }
    public required string Title { get; set; }

    public required Trait MainTrait { get; set; }

    public required List<Trait> Traits { get; set; } // NOTE: Would no longer throw if you remove this line but then the `MainTrait` column won't be named correctly either.
}

class Trait
{
    public required string Title { get; set; }
    public required string Description { get; set; }
}
```
`.csproj`:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <RootNamespace>efcore_json_test</RootNamespace>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="EFCore.NamingConventions" Version="8.0.0-rc.2" />
    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="8.0.0" />
  </ItemGroup>

</Project>
```

Exception:
```
Unhandled exception. System.InvalidOperationException: Entity 'Product.Traits#Trait (Trait)' is mapped to JSON and also to a table or view 'products_traits', but its owner 'Product' is mapped to a different table or view 'products'. Every entity mapped to JSON must also map to the same table or view as its owner.
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.ValidateJsonEntities(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Infrastructure.Internal.NpgsqlModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelRuntimeInitializer.Initialize(IModel model, Boolean designTime, IDiagnosticsLogger`1 validationLogger)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, ModelCreationDependencies modelCreationDependencies, Boolean designTime)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel(Boolean designTime)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__8_4(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass2_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_ContextServices()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService(IInfrastructure`1 accessor, Type serviceType)
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_Dependencies()
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureDeleted()
   at Program.<Main>$(String[] args) in /home/arad/scratchpad/efcore-json-test/Program.cs:line 5
```

**Comment by markushaslinger** [comment]
@aradalvand I think one issue you have is trying to map a collection (`Traits`) to JSON - [it seems EF does not yet allow for collections to be turned into arrays and one has to use a 'wrapper' object](https://stackoverflow.com/questions/77207781/entity-framework-json-column-mapping) (e.g. a `TraitCollection` which can be mapped using `OwnsOne` and itself contains the `List<Trait>`). At least that is what worked for me.

@roji To add some information to the concrete issue: I got it to work with JSON columns as described above, but in the end _the naming convention for the JSON column was not applied_ (so instead of `additional_data` it was named `AdditionalData`). It works, but having one column break the naming convention is weird also.
I tried to manually set a column name (to what it should be) using `HasColumnName`, but then I got an exception like the above.

**Comment by roji** [comment]
Thanks for the additional info @markushaslinger, I'll make sure to test that this stuff works when I get around to this issue.

**Comment by aradalvand** [comment]
> @aradalvand I think one issue you have is trying to map a collection (Traits) to JSON - [it seems EF does not yet allow for collections to be turned into arrays and one has to use a 'wrapper' object](https://stackoverflow.com/questions/77207781/entity-framework-json-column-mapping) (e.g. a TraitCollection) which can be mapped using OwnsOne and itself contains the List<Trait>.

That is absolutely not the case. I literally mentioned in the original code snippet that everything starts working if you just remove the call to `UseSnakeCaseNamingConvention`. So, EF does very much support this type of scenario.

> @roji To add some information to the concrete issue: I got it to work with JSON columns as described above, but in the end the naming convention for the JSON column was not applied (so instead of additional_data it was named AdditionalData).

Again, I literally mentioned that in the original code snippet (as a comment), you've provided no additional useful information here.

**Comment by xamir82** [comment]
Is there a workaround for this one analogous to [this](https://github.com/efcore/EFCore.NamingConventions/issues/184#issuecomment-1386783955)?

**Comment by moxplod** [comment]
@xamir82 No, I have been facing this issue and am unable to use the new JSON features of EF Core 7+ due to this. 

I have been decorating the Property with [Column(TypeName = "jsob")] in NPGSql. But this loses the new JSON features of querying. 

Here is my classes and error to add more context here. I use  .UseLowerCaseNamingConvention();
```

public class Activity 
{
... 

    public ActivityProps Properties { get; set; }

}

public class ActivityProps
{
    public int? Rating { get; set; }
}

```

The error I get is: 

> Unable to create a 'DbContext' of type ''. The exception 'Entity 'ActivityProps' is mapped to JSON and also to a table or view 'activityprops', but its owner 'Activity' is mapped to a different table or view 'activity'. Every entity mapped to JSON must also map to the same table or view as its owner.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728

**Comment by moxplod** [comment]
Hi, I thought this would fix my issue, but I just upgraded to 8.0.3 and I still get this error : 

> Unable to create a 'DbContext' of type ''. The exception 'Entity 'ActivityProps' is mapped to JSON and also to a table or view 'activityprops', but its owner 'Activity' is mapped to a different table or view 'activity'. Every entity mapped to JSON must also map to the same table or view as its owner.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728


I am using UseLowerCaseNamingConvention. That is why, in the error, you can notice the comparison of ActivityProps to lowercase activityprops. 

Here are the relevant snippets of code:

```
modelBuilder.Entity<Activity>().OwnsOne(a => a.Properties, 
            b => b.ToJson());

public class Activity 
{
... 

    public ActivityProps Properties { get; set; }

}

public class ActivityProps
{
    public int? Rating { get; set; }
}

```

**Comment by roji** [comment]
@moxplod can you please open a new issue with a minimal, runnable repro?

**Comment by GalensGan** [comment]
> Hi, I thought this would fix my issue, but I just upgraded to 8.0.3 and I still get this error :
> 
> > Unable to create a 'DbContext' of type ''. The exception 'Entity 'ActivityProps' is mapped to JSON and also to a table or view 'activityprops', but its owner 'Activity' is mapped to a different table or view 'activity'. Every entity mapped to JSON must also map to the same table or view as its owner.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728
> 
> I am using UseLowerCaseNamingConvention. That is why, in the error, you can notice the comparison of ActivityProps to lowercase activityprops.
> 
> Here are the relevant snippets of code:
> 
> ```
> modelBuilder.Entity<Activity>().OwnsOne(a => a.Properties, 
>             b => b.ToJson());
> 
> public class Activity 
> {
> ... 
> 
>     public ActivityProps Properties { get; set; }
> 
> }
> 
> public class ActivityProps
> {
>     public int? Rating { get; set; }
> }
> ```

same issue for me, is there some iedea?

**Comment by gerfen** [comment]
Not sure if this will help anyone but I'm applying a custom table convention and got a similar error.  I was able to work around the issue by adding `[NotMapped]` to the offending Entity.  I'm using 8.0.5.

For example:

```
    public abstract class TokenComponent : IdentifiableEntity
    {
      ...

        public virtual List<Metadatum> Metadata { get; set; } = new();
    }

    [NotMapped]
    public class Metadatum
    {
        public string Key { get; set; }

       public string? Value { get; set; }
    }
```
